File: ausf/src/did-proxy/registration/bcf_client.go
/*
 * Licensed to the OpenAirInterface (OAI) Software Alliance under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The OpenAirInterface Software Alliance licenses this file to You under
 * the OAI Public License, Version 1.1  (the "License"); you may not use this
 * file except in compliance with the License. You may obtain a copy of the
 * License at
 *
 *      http://www.openairinterface.org/?page_id=698
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *-------------------------------------------------------------------------------
 * For more information about the OpenAirInterface (OAI) Software Alliance:
 *      contact@openairinterface.org
 */

package registration

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/oai/did-proxy/config"
	"github.com/oai/did-proxy/logger"
	"github.com/oai/did-proxy/profile"
)

// BCFClient handles communication with the BCF (Blockchain Function)
// Following the same pattern as NRF registration in OAI AUSF
type BCFClient struct {
	bcfConfig   *config.BCFConfig
	httpVersion int
	httpClient  *http.Client
	log         *logger.Logger
	registered  bool
}

// NewBCFClient creates a new BCF registration client
func NewBCFClient(bcfCfg *config.BCFConfig, httpVersion int, log *logger.Logger) *BCFClient {
	// Create HTTP client with timeout
	client := &http.Client{
		Timeout: 30 * time.Second,
		Transport: &http.Transport{
			MaxIdleConns:        10,
			MaxIdleConnsPerHost: 10,
			IdleConnTimeout:     90 * time.Second,
		},
	}

	return &BCFClient{
		bcfConfig:   bcfCfg,
		httpVersion: httpVersion,
		httpClient:  client,
		log:         log,
		registered:  false,
	}
}

// RegisterNFProfile registers the extended NF profile with BCF
// Similar to NF registration with NRF (PUT request)
func (c *BCFClient) RegisterNFProfile(extProfile *profile.ExtendedNFProfile) error {
	c.log.Info("Sending NF Instance Registration to BCF")

	uri := fmt.Sprintf("%s/%s", c.bcfConfig.GetURI(), extProfile.NFInstanceID)
	c.log.Debug("BCF Registration URI: %s", uri)

	// Prepare request body
	body := extProfile.ToBytes()
	c.log.Debug("Request body size: %d bytes", len(body))

	// Create PUT request (same as NRF registration)
	req, err := http.NewRequest(http.MethodPut, uri, bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")

	// Send request
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	// Read response
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("failed to read response: %w", err)
	}

	// Check response status (CREATED or OK indicates success)
	if resp.StatusCode == http.StatusCreated || resp.StatusCode == http.StatusOK {
		c.log.Info("NFRegistration successful, got response from BCF (status: %d)", resp.StatusCode)
		c.log.Debug("Response body: %s", string(respBody))
		c.registered = true
		return nil
	}

	// Handle error response
	c.log.Error("BCF registration failed with status: %d", resp.StatusCode)
	c.log.Debug("Error response: %s", string(respBody))
	return fmt.Errorf("BCF registration failed with status %d: %s", resp.StatusCode, string(respBody))
}

// DeregisterNFProfile deregisters the NF from BCF
// Similar to NF deregistration from NRF (DELETE request)
func (c *BCFClient) DeregisterNFProfile(nfInstanceID string) error {
	if !c.registered {
		c.log.Debug("NF was not registered, skipping deregistration")
		return nil
	}

	c.log.Info("Sending NF Instance Deregistration to BCF")

	uri := fmt.Sprintf("%s/%s", c.bcfConfig.GetURI(), nfInstanceID)
	c.log.Debug("BCF Deregistration URI: %s", uri)

	// Create DELETE request
	req, err := http.NewRequest(http.MethodDelete, uri, nil)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Accept", "application/json")

	// Send request
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	// Check response status (NO_CONTENT indicates success)
	if resp.StatusCode == http.StatusNoContent || resp.StatusCode == http.StatusOK {
		c.log.Info("NF Deregistration successful from BCF")
		c.registered = false
		return nil
	}

	return fmt.Errorf("BCF deregistration failed with status %d", resp.StatusCode)
}

// UpdateNFProfile sends a heartbeat/update to BCF
// Similar to NRF heartbeat (PATCH request)
func (c *BCFClient) UpdateNFProfile(nfInstanceID string, patchItems []PatchItem) error {
	c.log.Debug("Sending NF Update to BCF")

	uri := fmt.Sprintf("%s/%s", c.bcfConfig.GetURI(), nfInstanceID)
	c.log.Debug("BCF Update URI: %s", uri)

	// Prepare PATCH body
	body, err := json.Marshal(patchItems)
	if err != nil {
		return fmt.Errorf("failed to marshal patch items: %w", err)
	}

	// Create PATCH request
	req, err := http.NewRequest(http.MethodPatch, uri, bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json-patch+json")
	req.Header.Set("Accept", "application/json")

	// Send request
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusNoContent {
		c.log.Debug("NF Update successful")
		return nil
	}

	return fmt.Errorf("BCF update failed with status %d", resp.StatusCode)
}

// Heartbeat sends a periodic heartbeat to BCF
func (c *BCFClient) Heartbeat(nfInstanceID string) error {
	patchItems := []PatchItem{
		{
			Op:    "replace",
			Path:  "/nfStatus",
			Value: "REGISTERED",
		},
	}
	return c.UpdateNFProfile(nfInstanceID, patchItems)
}

// IsRegistered returns whether the NF is currently registered with BCF
func (c *BCFClient) IsRegistered() bool {
	return c.registered
}

// PatchItem represents a JSON Patch item
type PatchItem struct {
	Op    string      `json:"op"`
	Path  string      `json:"path"`
	Value interface{} `json:"value,omitempty"`
	From  string      `json:"from,omitempty"`
}

// RegistrationResponse represents the BCF registration response
type RegistrationResponse struct {
	NFInstanceID   string                 `json:"nfInstanceId"`
	NFStatus       string                 `json:"nfStatus"`
	HeartBeatTimer int                    `json:"heartBeatTimer,omitempty"`
	CustomInfo     map[string]interface{} `json:"customInfo,omitempty"`
}

File: amf/src/bcf/bcf_client.cpp
/*
 * Licensed to the OpenAirInterface (OAI) Software Alliance under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The OpenAirInterface Software Alliance licenses this file to You under
 * the OAI Public License, Version 1.1  (the "License"); you may not use this
 * file except in compliance with the License. You may obtain a copy of the
 * License at
 *
 *      http://www.openairinterface.org/?page_id=698
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *-------------------------------------------------------------------------------
 * For more information about the OpenAirInterface (OAI) Software Alliance:
 *      contact@openairinterface.org
 */

#include "bcf_client.hpp"

#include <curl/curl.h>

#include <cstring>

#include "logger.hpp"

namespace oai::common::bcf {

//------------------------------------------------------------------------------
// CURL 回调函数
//------------------------------------------------------------------------------

static size_t write_callback(
    void* contents, size_t size, size_t nmemb, void* userp) {
  size_t total_size = size * nmemb;
  std::string* str  = static_cast<std::string*>(userp);
  str->append(static_cast<char*>(contents), total_size);
  return total_size;
}

//------------------------------------------------------------------------------
// BcfHttpClient 实现
//------------------------------------------------------------------------------

BcfHttpClient::BcfHttpClient()
    : m_http_version(2), m_timeout_ms(5000) {}

void BcfHttpClient::set_base_uri(const std::string& uri) {
  m_base_uri = uri;
  // 移除尾部斜杠
  while (!m_base_uri.empty() && m_base_uri.back() == '/') {
    m_base_uri.pop_back();
  }
}

void BcfHttpClient::set_http_version(int version) {
  m_http_version = version;
}

void BcfHttpClient::set_timeout(uint32_t timeout_ms) {
  m_timeout_ms = timeout_ms;
}

HttpResponse BcfHttpClient::get(const std::string& path) {
  return send_request(HttpMethod::GET, path);
}

HttpResponse BcfHttpClient::post(
    const std::string& path, const std::string& body) {
  return send_request(HttpMethod::POST, path, body);
}

HttpResponse BcfHttpClient::post(
    const std::string& path, const nlohmann::json& json_body) {
  return post(path, json_body.dump());
}

HttpResponse BcfHttpClient::put(
    const std::string& path, const std::string& body) {
  return send_request(HttpMethod::PUT, path, body);
}

HttpResponse BcfHttpClient::put(
    const std::string& path, const nlohmann::json& json_body) {
  return put(path, json_body.dump());
}

HttpResponse BcfHttpClient::del(const std::string& path) {
  return send_request(HttpMethod::DELETE, path);
}

HttpResponse BcfHttpClient::send_request(
    HttpMethod method, const std::string& path, const std::string& body) {
  std::string full_uri = m_base_uri + path;
  return do_send(http_method_to_string(method), full_uri, body);
}

HttpResponse BcfHttpClient::do_send(
    const std::string& method, const std::string& full_uri,
    const std::string& body) {
  HttpResponse response;

  CURL* curl = curl_easy_init();
  if (!curl) {
    response.error_message = "Failed to initialize CURL";
    return response;
  }

  std::string response_body;

  // 设置 URL
  curl_easy_setopt(curl, CURLOPT_URL, full_uri.c_str());

  // 设置 HTTP 版本
  if (m_http_version == 2) {
    curl_easy_setopt(curl, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_2_0);
  } else {
    curl_easy_setopt(curl, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);
  }

  // 设置超时
  curl_easy_setopt(curl, CURLOPT_TIMEOUT_MS, m_timeout_ms);
  curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT_MS, m_timeout_ms / 2);

  // 设置回调
  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
  curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response_body);

  // 设置 HTTP 方法
  if (method == "GET") {
    curl_easy_setopt(curl, CURLOPT_HTTPGET, 1L);
  } else if (method == "POST") {
    curl_easy_setopt(curl, CURLOPT_POST, 1L);
  } else if (method == "PUT") {
    curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "PUT");
  } else if (method == "DELETE") {
    curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "DELETE");
  } else if (method == "PATCH") {
    curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "PATCH");
  }

  // 设置请求体
  if (!body.empty()) {
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, body.c_str());
    curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, body.size());
  }

  // 设置请求头
  struct curl_slist* headers = nullptr;
  headers = curl_slist_append(headers, "Content-Type: application/json");
  headers = curl_slist_append(headers, "Accept: application/json");
  curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

  // 忽略 SSL 证书验证 (开发环境)
  curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0L);
  curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 0L);

  // 执行请求
  CURLcode res = curl_easy_perform(curl);

  if (res == CURLE_OK) {
    long http_code = 0;
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);

    response.status_code = static_cast<uint32_t>(http_code);
    response.body        = response_body;
    response.success     = response.is_success();
    response.parse_json();
  } else {
    response.error_message = curl_easy_strerror(res);
  }

  // 清理
  curl_slist_free_all(headers);
  curl_easy_cleanup(curl);

  return response;
}

}  // namespace oai::common::bcf
